### PR TITLE
feat: Compute Top Performing Students

### PR DESCRIPTION
**Pull request by wais-hundekar**
## Notion ticket link

[Compute Top Performing Students](https://www.notion.so/uwblueprintexecs/Task-Board-db95cd7b93f245f78ee85e3a8a6a316d)

## Implementation description
* Implemented resolver, function and query to compute the top 5 students for a given testSessionId
Screenshot 2023-07-08 at 3 54 11 PM

## Steps to test
1. Ensure existing tests are accurate
2. Call function on test sessions within the backend, (ensure expired date has passed)

## What should reviewers focus on?
* 

## Checklist
- [ ] My PR name is descriptive and in imperative tense
- [ ] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [ ] I have run the appropriate linter(s)
- [ ] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/graphql/resolvers/testSessionResolvers.ts [hunk 1]
```diff
@@ -41,6 +41,10 @@ const testSessionResolvers = {
       _parent: undefined,
       { teacherId }: { teacherId: string },
     ) => testSessionService.getTestSessionsByTeacherId(teacherId),
+    getTop5StudentsByTestSessionId: (
```
Comment (carissa-tang) [inline backend/graphql/resolvers/testSessionResolvers.ts]:
```suggestion
    getTopFiveStudentsById: (
```
nit: rename function

File: backend/services/interfaces/testSessionService.ts [hunk 2]
```diff
@@ -212,4 +212,11 @@ export interface ITestSessionService {
     id: string,
     result: ResultRequestDTO,
   ): Promise<TestSessionResponseDTO>;
+
+  /**
+   * Get the top 5 best students for a test session
```
Comment (carissa-tang) [inline backend/services/interfaces/testSessionService.ts]:
```suggestion
   * Get the top 5 students (based on test score) for a test session
```

File: frontend/src/APIClients/queries/TestSessionQueries.ts [hunk 3]
```diff
@@ -89,3 +89,9 @@ export const GET_TEST_SESSION_TITLE = gql`
     }
   }
 `;
+
+export const GET_TOP_5_STUDENTS_BY_TEST_SESSION_ID = gql`
+  query Query($getTop5StudentsByTestSessionIdId: ID!) {
+    getTop5StudentsByTestSessionId(id: $getTop5StudentsByTestSessionIdId)
```
Comment (carissa-tang) [inline frontend/src/APIClients/queries/TestSessionQueries.ts]:
```suggestion
  query Query($id: ID!) {
    getTop5StudentsByTestSessionId(id: $id)
```

File: backend/services/implementations/testSessionService.ts [hunk 4]
```diff
@@ -502,6 +502,39 @@ class TestSessionService implements ITestSessionService {
...
+        testSessionId,
+      );
+
+      // Check if the endDate for the test session has passed
+      const currentDate = new Date();
+      if (testSession.endDate > currentDate) {
+        throw new Error("Test session has not ended yet");
+      }
+
+      const results = testSession.results || [];
```
Comment (carissa-tang) [inline backend/services/implementations/testSessionService.ts]: can you add types to all these variables?

File: backend/services/implementations/testSessionService.ts [hunk 5]
```diff
@@ -502,6 +502,39 @@ class TestSessionService implements ITestSessionService {
...
+
+      // Sort the results based on the score in descending order
+      const sortedResults = results.sort((a, b) => b.score - a.score);
+
+      // Get the top 5 students' IDs
+      const top5Students = sortedResults
```
Comment (carissa-tang) [inline backend/services/implementations/testSessionService.ts]:
```suggestion
      const topFiveStudents = sortedResults
```

**Review by carissa-tang — CHANGES_REQUESTED** [review]

File: frontend/src/APIClients/queries/TestSessionQueries.ts [hunk 6]
```diff
@@ -89,3 +89,9 @@ export const GET_TEST_SESSION_TITLE = gql`
     }
   }
 `;
+
+export const GET_TOP_5_STUDENTS_BY_TEST_SESSION_ID = gql`
+  query Query($id: ID!) {
+    gettop5StudentsById(id: $id)
```
Comment (carissa-tang) [inline frontend/src/APIClients/queries/TestSessionQueries.ts]:
```suggestion
    getTopFiveStudentsById(id: $id)
```

File: backend/services/implementations/testSessionService.ts [hunk 7]
```diff
@@ -502,6 +502,39 @@ class TestSessionService implements ITestSessionService {
...
+  async getTopFiveStudentsById(testSessionId: string): Promise<Array<string>> {
+    try {
+      const testSession: TestSessionResponseDTO = await this.getTestSessionById(
+        testSessionId,
+      );
+
+      // Check if the endDate for the test session has passed
+      const currentDate: Date = new Date();
+      if (testSession.endDate > currentDate) {
+        throw new Error("Test session has not ended yet");
```
Comment (carissa-tang) [inline backend/services/implementations/testSessionService.ts]: can we add the `testSessionId` into the error message?

File: backend/services/implementations/__tests__/testSessionService.test.ts [hunk 8]
```diff
@@ -458,4 +459,55 @@ describe("mongo testSessionService", (): void => {
...
+    );
+
+    expect(topStudents).toEqual([]);
+  });
+
+  it("getTopFiveStudentsById with an error retrieving test session", async () => {
+    const testSessionId = "62c248c0f79d6c3c9ebbea92";
+
+    // Create a mock implementation for getTestSessionById that throws an error
+    testSessionService.getTestSessionById = jest
```
Comment (carissa-tang) [inline backend/services/implementations/__tests__/testSessionService.test.ts]: why do you need to mock this?

**Review by carissa-tang — COMMENTED** [review]
few more comments

File: backend/services/implementations/__tests__/testSessionService.test.ts (hunk 8 above)
Comment (wais-hundekar) [reply backend/services/implementations/__tests__/testSessionService.test.ts]: good point!

File: backend/services/implementations/__tests__/testSessionService.test.ts [hunk 9]
```diff
@@ -458,4 +459,54 @@ describe("mongo testSessionService", (): void => {
...
+    );
+
+    expect(topStudents).toEqual([]);
+  });
+
+  it("getTopFiveStudentsById with an error retrieving test session", async () => {
+    const testSessionId = "62c248c0f79d6c3c8ebbea92";
+
+    // Create a mock implementation for getTestSessionById that throws an error
+
```
Comment (carissa-tang) [inline backend/services/implementations/__tests__/testSessionService.test.ts]:
```suggestion
```

File: backend/services/implementations/__tests__/testSessionService.test.ts [hunk 10]
```diff
@@ -458,4 +459,54 @@ describe("mongo testSessionService", (): void => {
...
+      "Test Session id 62c248c0f79d6c3c8ebbea92 not found",
+    );
+  });
+
+  it("should throw an error if the test session's end date has not passed", async () => {
+    const savedTestSession = await MgTestSession.create(mockTestSession);
+
+    await expect(
+      testSessionService.getTopFiveStudentsById(savedTestSession.id),
+    ).rejects.toThrowError("Test session has not ended yet");
```
Comment (carissa-tang) [inline backend/services/implementations/__tests__/testSessionService.test.ts]:
```suggestion
    ).rejects.toThrowError(`Test session has not ended yet. testSessionId: ${savedTestSession.id}`);
```

**Review by carissa-tang — APPROVED** [review]
minor nits, but lgtm!

**Pull request merged**